wallet/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+    utxos: rpds::HashTrieMapSync<NoteId, Utxo>,
+    pk_index: rpds::HashTrieMapSync<PublicKey, rpds::HashTrieSetSync<NoteId>>,
+}
+
+impl WalletState {
+    pub fn from_ledger(known_keys: &HashSet<PublicKey>, ledger: &LedgerState) -> Self {
+        let mut utxos = rpds::HashTrieMapSync::new_sync();
+        let mut pk_index = rpds::HashTrieMapSync::new_sync();
+
+        for (_, (utxo, _)) in ledger.latest_commitments().utxos().iter() {
+            if known_keys.contains(&utxo.note.pk) {
+                let note_id = utxo.id();
+                utxos = utxos.insert(note_id, *utxo);
+
+                let note_set = pk_index
+                    .get(&utxo.note.pk)
+                    .cloned()
+                    .unwrap_or_else(rpds::HashTrieSetSync::new_sync)
+                    .insert(note_id);
+                pk_index = pk_index.insert(utxo.note.pk, note_set);
+            }
+        }
+
+        Self { utxos, pk_index }
+    }
+
+    pub fn utxos_for_amount(
+        &self,
+        amount: Value,
+        pks: impl IntoIterator<Item = PublicKey>,
+    ) -> Option<Vec<Utxo>> {
+        let mut utxos: Vec<Utxo> = pks
+            .into_iter()
+            .filter_map(|pk| self.pk_index.get(&pk))
+            .flatten()
+            .map(|id| self.utxos[id])
+            .collect();
+
+        // we want to consume small valued notes first to keep our wallet tidy
+        utxos.sort_by_key(|utxo| utxo.note.value);
+
+        let mut selected_utxos = Vec::new();
+        let mut selected_amount = 0;
+
+        for utxo in utxos {
+            selected_utxos.push(utxo);
+            selected_amount += utxo.note.value;
+            if selected_amount >= amount {
+                break;
+            }
+        }
+
+        if selected_amount < amount {
+            None
+        } else {
+            // We may have smaller notes that are redundant.
+            //
+            // e.g. Suppose we hold notes valued [3 NMO, 4 NMO] and we asked for 4 NMO
+            //      then, since we sort the notes by value, we would have first added
+            //      the 3 NMO note and then then 4 NMO note to the selected utxos list.
+            //
+            //      The 4 NMO note alone would have satisfied the request, the 3 NMO note is
+            //      redundant and would be returned as change in a transaction.
+            //
+            // To resolve this, we remove as many of the smallest notes as we can while
+            // still keep us above the requested amount.
+
+            let mut skip_count = 0;
+            let mut temp_amount = selected_amount;
+
+            for utxo in &selected_utxos {
+                if temp_amount - utxo.note.value >= amount {
+                    temp_amount -= utxo.note.value;
+                    skip_count += 1;
+                } else {
+                    break;
+                }
+            }
+
+            selected_utxos.drain(..skip_count);
+
+            Some(selected_utxos)
```

Comment:
I think this should be its own method. The comment should be part of the method documentation.

Reply:
yes, that'd be better, done!